### ISSUE TITLE
Revamp compile error formatting & output

### ISSUE THREAD

**Issue by paulcsmith** [crystal-lang/crystal#7553]
Based on proposal in https://forum.crystal-lang.org/t/rfc-revamp-how-compile-time-errors-are-shown/548/3

## The Proposed Changes

* By default, just show the code with the error, not the entire trace. This is what Elm does and it works quite well. Helps focus on what is necessary.
* Add a command line option and/or ENV to show the full trace. Sometimes it is useful to see more, but let's make that optional instead of making the current verbose output the default. 
* Some changes to how the error is displayed to make it easier to scan at a glance. Code + pics attached

## Example Code and Screenshots

https://github.com/paulcsmith/crystal-errors

### For errors without more info (just one line)

* Prefix with "Error"
* Print error message in yellow and bold
* (not sure about this) show how to print the full trace (like it does now)
* Show what file it is
* Show line number right by the source code line
* Show code line (maybe bolded?) 

https://github.com/paulcsmith/crystal-errors/blob/master/errors/type_restriction.cr

<img alt="Screen Shot 2019-03-14 at 9 49 07 PM" src="https://user-images.githubusercontent.com/22394/54437417-f0ff0280-470a-11e9-9e2d-8503b61d9282.png">

### For errors with more than one line

* Print first line in yellow just like other errors
* **Everything after first newline goes after the source line.**

https://github.com/paulcsmith/crystal-errors/blob/master/errors/no_overload.cr

<img alt="Screen Shot 2019-03-14 at 9 48 37 PM" src="https://user-images.githubusercontent.com/22394/54437329-c44aeb00-470a-11e9-9df5-7746284b90b7.png">

### Macro errors

* Print surrounding lines
* Only bold line with the problem
* **Show the expanded code, and underneath it show the unexpanded code**

It is important to show both. The unexpanded part show what the user wrote. The expanded part helps figure out why the issue might be there. Use --verbose/--debug for more in-depth look if it still is not enough

https://github.com/paulcsmith/crystal-errors/blob/master/errors/macro_error.cr

<img alt="Screen Shot 2019-03-14 at 10 18 12 PM" src="https://user-images.githubusercontent.com/22394/54437336-c90f9f00-470a-11e9-8347-ff711b9aeef0.png">

### Flag name `--trace-error`

`crystal build src/my_app.cr --trace-error`

This will print all the frames. Surround error message with `--` and keep it yellow and bold to highlight the final frame and the error message

### Env var `TRACE_CRYSTAL_ERRORS`

It might be nice to allow setting this. That way if someone always/never wants to trace Crystal errors, they can control that by setting this var in their shell. 

This may need further discussion and can likely be done separately!

### If it turns out we need more frames to make this useful

Surround error message with `--` and keep it yellow so there is a clear delineation between the final frame and previous ones

### Similar issues

https://github.com/crystal-lang/crystal/issues/7554 make messages friendlier

**Comment by paulcsmith** [comment]
Bounty posted for whoever wants to tackle this ❤️ $125 as of now https://www.bountysource.com/issues/71440311-revamp-compile-error-output

**Comment by martimatix** [comment]
I started toying with this a week ago but didn't say anything at first because I had never worked on the Crystal compiler and didn't know how far I'd get. I'm pleased to report that I've got `SyntaxException`  and `TypeException` working for both macro and non-macro cases which covers most of the common errors.

Work left to do:

1. Wire up the `--full-trace` command line flag.
1. Address the `@color` instance variable in the exceptions - need to ensure that everything looks ok when `--no-color` flag is used.
1. Address specs.
1. Refactoring and clean up.

And then I should be able to open up a PR.

**Comment by paulcsmith** [comment]
This is incredible news!! Nice work @martimatix

**Comment by jwoertink** [comment]
Boom! This is gonna be huge @martimatix

**Comment by martimatix** [comment]
@paulcsmith With reference to the macro expansion example, if I understand it correctly, Line 14 is the first line of the expanded macro and Line 13 of `src/my_cool_macro.cr` would be `macro my_cool_macro`.

This works nicely for macros where the macro expansion maps nicely to the corresponding unexpanded source code. However, consider the example below.

```crystal
macro define_method(name, content)
  def {{name}}
    {% if content == 1 %}
      "one"
    {% elsif content == 2 %}
      this_will_cause_an_error
    {% else %}
      {{content}}
    {% end %}
  end
end

define_method bar, 2

bar #=> this line triggers the compilation error
```

In my current implementation, the error message looks like this:

```
Error: undefined local variable or method 'this_will_cause_an_error' for top-level
Showing last frame. Use --all-frames for full trace.

Expanded macro from src/my-crystal-sandbox.cr:

   2 |   def bar
   3 |
 > 4 |       this_will_cause_an_error
   5 |

In src/my-crystal-sandbox.cr:

 13 | define_method bar, 2
      ^
```

In this situation, the error message is not particularly helpful because the macro expansion doesn't have a one-to-one mapping with the unexpanded source code. It's indicating that the error is on Line 4 of the source file but this is not correct.

## Possible work arounds

### Give the line number of where the macro is defined in the source code

This would be most similar to how the Crystal compiler currently handles this situation. The expanded macro definition would start with line number 1.

```
Expanded macro from src/my-crystal-sandbox.cr: 1:

   1 |   def bar
   2 |
 > 3 |       this_will_cause_an_error
   4 |
```

### Head the expanded macro code with the origin of the macro definition.

Offset the source code line number by one space and macro expansion begins from Line 1. The source code line could be magenta for the colour version.

```
  1    macro define_method(name, content)
   1 |   def bar
   2 |
 > 3 |       this_will_cause_an_error
   4 |
```

**Comment by paulcsmith** [comment]
That's a really good point. I think in this case the line numbers by the expanded macro aren't helpful since they don't correspond to anything the user can inspect.

So with that said I think show `Expanded 'define_method' from src/my-crystal-sandbox.cr:1` (note I changed "macro" to the name of the macro) makes sense and that the line numbers next to the expanded macro can be removed since they aren't helpful. I think it still makes sense to highlight the line that caused the issue with color though so it is clear what the problem is.

Does that make sense? What do you think?

```
Expanded 'define_method' from src/my-file.cr:1

     def bar
         
         this_will_cause_an_error

In src/my-crystal-sandbox.cr:

 13 | define_method bar, 2
```

**Comment by martimatix** [comment]
Regarding line numbers being unhelpful, it's worth noting that the Crystal compiler currently shows line numbers when expanding macros. To me, the value of line numbers is that it provides a visual anchor when one is debugging an erroneous expanded macro. Furthermore, in a pair programming situation, I can imagine one programmer turning to another and saying, "See, there's a problem on Line 3," without needing to point to the screen.

A note on the "Head the expanded macro code with the origin of the macro definition" from my earlier comment. These new error messages train the user to look for line numbers on the left side of the screen. Although a bit unusual, my intent was to provide a more consistent user experience by keeping the macro definition line number also on the left.

**Comment by paulcsmith** [comment]
Great points on all counts! Maybe it was just the formatting of the "Head the expanded macro code with the origin of the macro definition" that was throwing me off. Thoughts on this? 

```
In src/my-crystal-sandbox.cr:

 13 | define_method bar, 2

Which called a macro in src/my-file.cr

 1  | macro define_methods(name, content)

Which expanded to:

 1  |  def bar
 2  |      
 3  |    this_will_cause_an_error # This is colorized
```

It is a bit longer, but I think much clearer. It also switches the user code to the top (I think that is more intuitive, thoughts?)

**Comment by paulcsmith** [comment]
Also, were you able to shrink the expanded macro code so it just shows a few lines before/after? If not then I think this won't work because some macros expand to hundreds of lines and then it is super hard to find the code that called it. But if it is possible to only show a few lines around the macro then this would be great I think

**Comment by martimatix** [comment]
Cool! I like this.

Regarding showing a few lines around the macro, that is possible but there's a problem -- what if the error occurs on Line 100 of the macro but it's not possible to figure out the cause of the error unless you're able to see the expanded code on Line 2?

### Possible solutions

1. Show only a few lines but show everything on the `--all-frames`* flag. Of course, this would print all frames AND the expanded macro.
1. Introduce a new flag called `--full-macro`.

I have also considered halting printing of the macro after the line with the error has been printed. However, if one were using a macro to define a new Class, maybe the underlying problem is a method definition that is below the line where the error occurs. And this still wouldn't prevent the problem of printing hundreds of lines for an expanded macro.

*I've been using `--all-frames` lately because there's already a flag called `--error-trace` and thought it would get confusing if there were both `--full-trace` and `--error-trace`flags.

**Comment by paulcsmith** [comment]
@martimatix Good point! I think that keeping a few lines by default is good and allowing the option to show all frames + expanded macro def is the best bet. That way you can see a high level view and then dig in only if necessary. If we show everything up front it can make it really hard to even have a rough idea of what went wrong or what you did to mess it up.

Also maybe this order would make it so it works with the full macro context and w/o:

```
Macro  in src/my-file.cr

 1  | macro define_methods(name, content)

Was expanded to:

 1  |  def bar
 2  |      
 3  |    this_will_cause_an_error # This is colorized

By src/my-crystal-sandbox.cr:

 13 | define_method bar, 2
```

But I'm not sure. Whatever we choose will be better than what is there and can be iterated on later I think

**Comment by RX14** [comment]
I really dislike this approach, crystal shows backtraces on compile errors because context is *essential* to actually solving compile errors because of crystal's unique type system. In reducing the information displayed, you do not help people with compile errors, you just make them run the compiler with `--error-trace` always.

The correct solution to shorter compiler backtraces is to implement casting argument values to argument types, then you can cut off backtraces as soon as you arrive at a fully-typed function.

**Comment by paulcsmith** [comment]
Fine. The rest of the work is still quite fantastic I think. 

This PR can easily be changed to show the full trace by default and would still benefit greatly from all the other changes.

**Comment by martimatix** [comment]
In the wild, what percentage of Crystal compile time errors can be solved just by looking at the last frame?

In regards to the PR,  having `--error-trace` always on is easy but there's a bit of a clean up effort involved with this.

**Comment by jwoertink** [comment]
As someone that actively uses crystal in production, I'm willing to share some stack traces I get. Some favorites are ones that look like

```
Errno???:0
Error writing to socket: Connection reset by peer
/usr/share/crystal/src/thread.cr:69:3 in 'unbuffered_write'<file>
```

or this

```
Exception???:0
???<file>
src/components/sponsor_list_component.cr:10:51 in '->'<file>
```

I think the main thing though is, any thing that helps error message to be clearer, and speeds up development will be a huge win for this language.

**Comment by martimatix** [comment]
@jwoertink You got me worried now because I haven't come across messages like that during my development of the PR. Hopefully, everything play nice and there won't be any regressions.

**Comment by asterite** [comment]
Those are runtime traces. It's a different issue.

**Comment by martimatix** [comment]
Below is a screenshot that @paulcsmith attached to a comment in the PR. This is _without_ `--error-trace`.

<img src="https://user-images.githubusercontent.com/22394/58728134-3e0e7e80-83b4-11e9-97b2-df67e6d988a4.png"/>

The biggest difference between using `--error-trace` and not using it is the way that macros are displayed. In the screenshot above, the expanded macro beyond the error line (Line 2) is hidden. In addition, the macro call location is displayed at the bottom (Line 269).

When `--error-trace` is used, the fully expanded macro is displayed. It is not necessary to display the call location because it is shown in the previous frame.

My concern is that if the `--error-trace` toggle is removed, then this feature is lost. 

I take @RX14's concerns seriously. At the same time, I hold a belief that it's beneficial to have shorter messages and only use `--error-trace` when necessary. That's why I'm curious as to how often one would need the full error trace.

**Comment by rishavs** [comment]
For whats its worth, as a low-prog-knowledge user of Crystal, I have always struggled with the error messaging. I myself have never used the full traces. So as far as I am concerned, these changes are most welcome.

I remember that when I first started using Crystal I actually stopped using it for a week or more because I couldnt understand the error messages. Turns out I was actually trying to read through the stack top to bottom as a complete newbie and would give up somewhere in the middle. Someone in the chat told me that I should instead read the console text from bottom upwards, as the actual error messages are at the bottom.  -__-

**Comment by bew** [comment]
Maybe we should add a line at the top like ruby & python iirc, to tell that the errors should be read from bottom to top

**Comment by asterite** [comment]
Closed by #7748

**Comment by paulcsmith** [comment]
@martimatix I  just realized the $125 bounty was never claimed (at least that's what it looks like)! You did a ton of work and totally deserve it. You can claim it here: https://www.bountysource.com/issues/71440311-revamp-compile-error-formatting-output
